plugins/redis/redisasync.cpp

```diff
@@ -0,0 +1,870 @@
+    void getVoidPtrLenPair(ICodeContext * ctx, const char * key, size_t & valueLength, void * & value, const char * channel);
+    //set
+    template<class type> void set(ICodeContext * ctx, const char * key, type value, unsigned expire, const char * channel);
+    template<class type> void set(ICodeContext * ctx, const char * key, size32_t valueLength, const type * value, unsigned expire, const char * channel);
+
+    bool missThenLock(ICodeContext * ctx, KeyLock * lockObject);
+    static void assertContextErr(const redisAsyncContext * context);
+
+protected :
+    virtual void selectDb(ICodeContext * ctx);
+    virtual void assertOnError(const redisReply * reply, const char * _msg);
+    virtual void assertConnection();
+    void createAndAssertConnection(ICodeContext * ctx);
+    void assertRedisErr(int reply, const char * _msg);
+    void handleLockForGet(ICodeContext * ctx, const char * key, const char * channel, ReturnValue * retVal);
+    void handleLockForSet(ICodeContext * ctx, const char * key, const char * value, size_t size, const char * channel, unsigned expire);
+    void subscribe(const char * channel, StringAttr & value);
+    void unsubscribe(const char * channel);
+    void attachLibev();
+    bool lock(const char * key, const char * channel);
+    void handleLoop(struct ev_loop * evLoop, ev_tstamp timeout);
+
+    //callbacks
+    static void assertCallbackError(const redisAsyncContext * context, const redisReply * reply, const char * _msg);
+    static void connectCB(const redisAsyncContext *c, int status);
+    static void disconnectCB(const redisAsyncContext *c, int status);
+    static void subCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void pubCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void setCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void getCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void unsubCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void setLockCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void selectCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void timeoutCB(struct ev_loop * evLoop, ev_timer *w, int revents);
+
+protected :
+    redisAsyncContext * context;
+};
+
+class SubContainer : public AsyncConnection
+{
+public :
+    SubContainer(ICodeContext * ctx, const char * options, const char * _channel, unsigned __int64 _database);
+    SubContainer(ICodeContext * ctx, const char * options, const char * _channel, redisCallbackFn * _callback, unsigned __int64 _database);
+    ~SubContainer()
+    {
+        unsubscribe();
+    }
+
+    inline void setCB(redisCallbackFn * _callback) { callback = _callback; }
+    inline redisCallbackFn * getCB() { return callback; }
+    inline void setChannel(const char * _channel) { channel.set(_channel); }
```

Comment:
Having lots of get/set functions (or public members) suggests the code isn't encapsulated correctly.

However in this case I think quite a few of these functions aren't used, so should probably be deleted.